apps/faucet/src/utils/api.ts

```diff
@@ -0,0 +1,83 @@
+import {
+  ChallengeResponse,
+  Data,
+  SettingsResponse,
+  TransferResponse,
+} from "./types";
+
+enum Endpoint {
+  Settings = "/setting",
+  Challenge = "/challenge",
+  Transfer = "",
+}
+
+export class API {
+  constructor(protected readonly url: string) {}
+
+  /**
+   * Wrapper for fetch requests to handle ReadableStream response when errors are received from API
+   *
+   * @param {string} endpoint
+   * @param {RequestInit} options
+   *
+   * @returns Object
+   */
+  async request<T = unknown>(
+    endpoint: string,
+    options: RequestInit = { method: "GET" }
+  ): Promise<T> {
+    return (await fetch(new URL(`${this.url}${endpoint}`), {
+      ...options,
+    })
+      .then((response) => {
+        if (response.ok) {
+          return response.json();
+        }
+        const reader = response?.body?.getReader();
+        return reader
+          ?.read()
```

Comment:
Actually, I was trying to figure out why TypeScript doesn't catch that. I'm not 100% sure, but if you add a return type to the `then` callback, it will catch it:

```typescript
return (await fetch(new URL(`${this.url}${endpoint}`), {
  ...options,                                           
})                                                      
  .then((response): Promise<T> => {
```

```
TS2322: Type 'Promise<T> | undefined' is not assignable to type 'Promise<T>'.
Type 'undefined' is not assignable to type 'Promise<T>'.
```

Reply:
Thanks! I'll take a look. I think in the case it can return `undefined`, we should throw an exception so it can be caught in the app. I wanted to remove any case of `undefined` as it's very tedious

Reply:
I added a type for that error response, with `{ message, code }`, so a correctly parsed error should always be this type, and in the case of `undefined` (if it can't get the reader and parse the error response) it will instead throw an error